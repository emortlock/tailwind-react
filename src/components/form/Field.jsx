import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'

import { withTheme } from '../theme'
import { BaseComponent } from '../tailwind'
import { getUniqueID } from '../utils'

class Field extends PureComponent {
  constructor(props) {
    super(props)

    this.id = getUniqueID('field')
  }

  render() {
    const {
      theme,
      is,
      children,
      className,
      hasHelp,
      hasError,
      disabled,
      ...rest
    } = this.props

    const fieldProps = {
      inputId: `${this.id}-input`,
      helpId: hasHelp ? `${this.id}-help` : undefined,
      errorId: hasError ? `${this.id}-error` : undefined,
      invalid: hasError,
      disabled,
    }

    return (
      <BaseComponent
        is={is}
        id={this.id}
        m={{ b: theme.spacing.md }}
        maxW="sm"
        {...rest}
      >
        {React.Children.map(children, child =>
          React.cloneElement(child, { field: fieldProps }),
        )}
      </BaseComponent>
    )
  }
}

Field.propTypes = {
  theme: PropTypes.shape({}).isRequired,
  is: PropTypes.oneOfType([PropTypes.string, PropTypes.func, PropTypes.object]),
  children: PropTypes.node,
  className: PropTypes.string,
  hasHelp: PropTypes.bool,
  hasError: PropTypes.bool,
  disabled: PropTypes.bool,
}

Field.defaultProps = {
  is: 'div',
  children: undefined,
  className: undefined,
  hasHelp: false,
  hasError: false,
  disabled: false,
}

export { Field as component }
export default withTheme(Field)
