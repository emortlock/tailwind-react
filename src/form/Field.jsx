import React from 'react'
import PropTypes from 'prop-types'
import classnames from 'classnames'

import { withTheme } from '../theme'
import {
  getTailwindClassNames,
  tailwindProps,
  tailwindPropTypes,
} from '../tailwind'
import { filterProps, getUniqueID } from '../utils'

const Field = ({
  theme,
  is,
  children,
  className,
  hasHelp,
  hasError,
  disabled,
  ...rest
}) => {
  const Component = is
  const userClassNames = classnames(getTailwindClassNames(rest), className)

  const fieldProps = {
    inputId: getUniqueID('field-'),
    helpId: hasHelp ? getUniqueID('field-info-') : undefined,
    errorId: hasError ? getUniqueID('field-error-') : undefined,
    invalid: hasError,
    disabled,
  }

  return (
    <Component
      {...filterProps(rest, tailwindProps)}
      className={classnames(
        'max-w-sm',
        `mb-${theme.spacing.md}`,
        userClassNames,
      )}
    >
      {React.Children.map(children, child =>
        React.cloneElement(child, { field: fieldProps }),
      )}
    </Component>
  )
}

Field.propTypes = {
  theme: PropTypes.shape({}).isRequired,
  is: PropTypes.oneOfType([PropTypes.string, PropTypes.func, PropTypes.object]),
  children: PropTypes.node,
  className: PropTypes.string,
  hasHelp: PropTypes.bool,
  hasError: PropTypes.bool,
  disabled: PropTypes.bool,
  ...tailwindPropTypes,
}

Field.defaultProps = {
  is: 'div',
  children: undefined,
  className: undefined,
  hasHelp: false,
  hasError: false,
  disabled: false,
}

export { Field as component }
export default withTheme(Field)
