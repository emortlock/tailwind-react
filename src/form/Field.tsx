import type { BoxProps } from 'tailwind-react-primitives'
import type { WithTheme } from '../theme'

import React from 'react'

import { RawBox as Box } from 'tailwind-react-primitives'
import { withTheme } from '../theme'
import { getUniqueID } from '../utils'

export interface WithField {
  field?: {
    inputId?: string
    invalid?: boolean
    disabled?: boolean
    labelId?: string
    errorId?: string
    helpId?: string
  }
}

export interface FieldProps<E extends HTMLElement = HTMLDivElement>
  extends WithTheme,
    BoxProps<E> {
  hasHelp?: boolean
  hasError?: boolean
  disabled?: boolean
  optionList?: boolean
}

export const RawField = <E extends HTMLElement = HTMLDivElement>({
  id,
  theme,
  as,
  children,
  className,
  hasHelp,
  hasError,
  disabled,
  optionList,
  ...rest
}: FieldProps<E>) => {
  const [fieldId, setFieldId] = React.useState(id)

  React.useEffect(() => {
    if (fieldId) return

    setFieldId(getUniqueID('field'))
  })

  const fieldProps: WithField['field'] = {
    inputId: `${fieldId}-input`,
    helpId: hasHelp ? `${fieldId}-help` : undefined,
    errorId: hasError ? `${fieldId}-error` : undefined,
    labelId: optionList ? `${fieldId}-label` : undefined,
    invalid: hasError,
    disabled,
  }

  return (
    <Box<E>
      as={as}
      id={fieldId}
      m={{ b: theme.spacing.md }}
      maxW="sm"
      {...rest}
    >
      {React.Children.map(children, (child) =>
        React.isValidElement(child)
          ? React.cloneElement<WithField>(child, { field: fieldProps })
          : child,
      )}
    </Box>
  )
}

export const Field = withTheme<FieldProps>(RawField)
