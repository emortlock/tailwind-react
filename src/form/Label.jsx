import React from 'react'
import PropTypes from 'prop-types'
import classnames from 'classnames'

import { withTheme } from '../theme'
import {
  getTailwindClassNames,
  tailwindProps,
  tailwindPropTypes,
} from '../tailwind'
import { filterProps } from '../utils'

const Label = ({
  theme,
  is,
  field: { inputId, disabled },
  children,
  className,
  htmlFor,
  ...rest
}) => {
  const Component = is
  const userClassNames = classnames(getTailwindClassNames(rest), className)

  return (
    <Component
      {...filterProps(rest, [...tailwindProps, 'invalid'])}
      className={classnames(
        `mb-${theme.spacing.sm}`,
        'inline-block',
        disabled && 'opacity-50',
        userClassNames,
      )}
      htmlFor={inputId || htmlFor}
    >
      {children}
    </Component>
  )
}

Label.propTypes = {
  theme: PropTypes.shape({}).isRequired,
  is: PropTypes.oneOfType([PropTypes.string, PropTypes.func, PropTypes.object]),
  field: PropTypes.shape({
    inputId: PropTypes.string,
    disabled: PropTypes.bool,
  }),
  children: PropTypes.node,
  className: PropTypes.string,
  htmlFor: PropTypes.string,
  ...tailwindPropTypes,
}

Label.defaultProps = {
  is: 'label',
  field: {
    disabled: false,
  },
  children: undefined,
  className: undefined,
  htmlFor: undefined,
}

export { Label as component }
export default withTheme(Label)
