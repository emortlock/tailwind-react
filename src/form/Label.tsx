import type { TextProps } from 'tailwind-react-primitives'
import type { WithTheme } from '../theme'
import type { WithField } from './Field'

import React from 'react'

import { RawText as Text } from 'tailwind-react-primitives'
import { withTheme } from '../theme'

export interface LabelProps<E extends HTMLElement = HTMLLabelElement>
  extends WithTheme,
    WithField,
    TextProps<E> {
  hasHelp?: boolean
  hasError?: boolean
  disabled?: boolean
  optionList?: boolean
  htmlFor?: string
}

export const RawLabel = <E extends HTMLElement = HTMLLabelElement>({
  theme,
  as = 'label',
  id,
  field: {
    labelId = undefined,
    inputId = undefined,
    disabled = undefined,
  } = {},
  children,
  htmlFor,
  optionList,
  ...rest
}: LabelProps<E>) => (
  <Text<E>
    as={as}
    id={labelId || id}
    inlineBlock
    htmlFor={!optionList ? inputId || htmlFor : undefined}
    m={{ b: theme.spacing.sm }}
    opacity={disabled ? 50 : undefined}
    weight="bold"
    {...rest}
  >
    {children}
  </Text>
)

export const Label = withTheme<LabelProps>(RawLabel)
