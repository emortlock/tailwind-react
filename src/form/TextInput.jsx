import React from 'react'
import PropTypes from 'prop-types'
import classnames from 'classnames'

import { withTheme } from '../theme'
import {
  getTailwindClassNames,
  tailwindProps,
  tailwindPropTypes,
} from '../tailwind'
import { filterProps } from '../utils'

const TextInput = ({
  theme,
  is,
  field,
  children,
  className,
  id,
  name,
  type,
  disabled,
  readOnly,
  invalid,
  ...rest
}) => {
  const Component = is
  const userClassNames = classnames(getTailwindClassNames(rest), className)

  const describedBy = [field.errorId, field.helpId].filter(by => by)

  return (
    <Component
      {...filterProps(rest, tailwindProps)}
      className={classnames(
        theme.radius,
        theme.textColors.body,
        `px-${theme.spacing.md} py-${theme.spacing.sm}`,
        `mb-${theme.spacing.sm}`,
        'appearance-none border w-full leading-tight',
        'focus:outline-none focus:shadow-outline',
        (readOnly || disabled) && 'opacity-50 pointer-events-none',
        invalid && `border-${theme.brandColors.danger}`,
        userClassNames,
      )}
      id={field.inputId || id || name}
      name={name}
      type={type}
      disabled={field.disabled || disabled}
      readOnly={readOnly}
      aria-invalid={field.invalid || invalid}
      aria-describedby={describedBy.length ? describedBy.join(' ') : undefined}
    />
  )
}

TextInput.propTypes = {
  theme: PropTypes.shape({}).isRequired,
  is: PropTypes.oneOfType([PropTypes.string, PropTypes.func, PropTypes.object]),
  field: PropTypes.shape({
    inputId: PropTypes.string,
    invalid: PropTypes.bool,
    disabled: PropTypes.bool,
  }),
  children: PropTypes.node,
  className: PropTypes.string,
  id: PropTypes.string,
  name: PropTypes.string.isRequired,
  type: PropTypes.string,
  disabled: PropTypes.bool,
  readOnly: PropTypes.bool,
  invalid: PropTypes.bool,
  ...tailwindPropTypes,
}

TextInput.defaultProps = {
  is: 'input',
  field: {},
  children: undefined,
  className: undefined,
  id: undefined,
  type: 'text',
  disabled: false,
  readOnly: false,
  invalid: false,
}

export { TextInput as component }
export default withTheme(TextInput)
