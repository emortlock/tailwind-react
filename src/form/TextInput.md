```jsx
import { Field, OptionLabel, Label, TextInput } from 'tailwind-react-ui'
;<>
  <Field>
    <Label>Name</Label>
    <TextInput name="name" />
  </Field>
</>
```
