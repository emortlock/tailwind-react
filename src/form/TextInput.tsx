import type { TouchableProps } from 'tailwind-react-primitives'
import type { WithTheme } from '../theme'
import type { WithField } from './Field'

import React from 'react'

import { RawTouchable as Touchable } from 'tailwind-react-primitives'
import { withTheme } from '../theme'

export interface TextInputProps<E extends HTMLElement = HTMLInputElement>
  extends WithTheme,
    WithField,
    TouchableProps<E> {
  invalid?: boolean
}

export const RawTextInput = <E extends HTMLElement = HTMLInputElement>({
  theme,
  as,
  field = {},
  children,
  id,
  name,
  type = 'text',
  disabled,
  readOnly,
  invalid,
  ...rest
}: TextInputProps<E>) => {
  const describedBy = [field.errorId, field.helpId].filter((by) => by)
  const isInvalid = field.invalid || invalid

  return (
    <Touchable<E>
      as={as}
      appearance="none"
      bg="white"
      rounded={theme.radius}
      text={theme.textColors.body}
      p={{ x: theme.spacing.md, y: theme.spacing.sm }}
      m={{ b: theme.spacing.sm }}
      border={!isInvalid ? true : [true, theme.brandColors.danger]}
      w="full"
      leading="tight"
      id={field.inputId || id || name}
      name={name}
      type={type}
      disabled={field.disabled || disabled}
      readOnly={readOnly}
      aria-invalid={isInvalid || undefined}
      aria-describedby={describedBy.length ? describedBy.join(' ') : undefined}
      {...rest}
    />
  )
}

export const TextInput = withTheme<TextInputProps>(RawTextInput)
