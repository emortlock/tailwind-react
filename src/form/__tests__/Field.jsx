import React from 'react'
import { shallow } from 'enzyme'

import { component as Field } from '../Field'
import { defaultConfig } from '../../config'

import Label from '../Label'
import TextInput from '../TextInput'
import InfoText from '../../typography/InfoText'
import DangerText from '../../typography/DangerText'

const setup = (testProps = {}) => {
  const props = Object.assign({ config: defaultConfig }, testProps)

  const wrapper = shallow(
    <Field {...props}>
      <Label>Password</Label>
      <InfoText>
        Your password must be 8-20 characters long, contain letters and numbers,
        and must not contain spaces, special characters, or emoji.
      </InfoText>
      <TextInput name="password" type="password" />
      <DangerText alert>Please complete</DangerText>
    </Field>,
  )

  return {
    props,
    wrapper,
  }
}

describe('Field', () => {
  it('renders matching snapshot', () => {
    const { wrapper } = setup()

    expect(wrapper).toMatchSnapshot()
  })
})
