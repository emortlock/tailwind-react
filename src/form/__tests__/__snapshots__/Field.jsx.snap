// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Field renders matching snapshot 1`] = `
<BaseComponent
  className="max-w-sm"
  is="div"
  m={
    Object {
      "b": 4,
    }
  }
>
  <WithTheme(Label)
    field={
      Object {
        "disabled": false,
        "errorId": "field-error-3",
        "helpId": "field-info-2",
        "inputId": "field-1",
        "invalid": true,
      }
    }
    key=".0"
  >
    Password
  </WithTheme(Label)>
  <HelpText
    field={
      Object {
        "disabled": false,
        "errorId": "field-error-3",
        "helpId": "field-info-2",
        "inputId": "field-1",
        "invalid": true,
      }
    }
    key=".1"
  >
    Your password must be 8-20 characters long, contain letters and numbers, and must not contain spaces, special characters, or emoji.
  </HelpText>
  <WithTheme(TextInput)
    field={
      Object {
        "disabled": false,
        "errorId": "field-error-3",
        "helpId": "field-info-2",
        "inputId": "field-1",
        "invalid": true,
      }
    }
    key=".2"
    name="password"
    type="password"
  />
  <ErrorText
    field={
      Object {
        "disabled": false,
        "errorId": "field-error-3",
        "helpId": "field-info-2",
        "inputId": "field-1",
        "invalid": true,
      }
    }
    key=".3"
  >
    Please complete
  </ErrorText>
</BaseComponent>
`;
